Openstax update

TylerZeroMaster opened the pull request:
TODO:
- [x] Make server port public in Gitpod or figure out another way to make it work
- [x] Add endpoints to server for additional openstax metadata
- [x] Save openstax metadata alongside h5p.json and content.json in a file named metadata.json
- [x] Add openstax metadata form to h5p editor
- [x] Validate form inputs before they are sent (98% Need some more details)
- [x] Private solutions and tips saved separately and inserted back into the content.json during editing
- [x] Better error handling (the h5p editor error handling is currently broken)
- [x] Tests
- [ ] Settings
- [ ] ...?

TylerZeroMaster committed:
Convert H5P custom editor into a panel

Make the panel wait for the server to start before appearing.
This should probably have a timeout attached to it as well.

Co-authored-by: Christopher M Kline <ckline-tryptic@users.noreply.github.com>

TylerZeroMaster committed:
Get everything to work as a bundle vsix

There are some additional json files required by the server that are
copied in the postbuild step. It is possible to get
rollup to bundle json files, maybe esbuild can too.

We also needed to change the way workspace root is passed into the server.

TylerZeroMaster committed:
Add configuration class for server

TylerZeroMaster committed:
Create form and server for openstax metadata

TylerZeroMaster committed:
Begin adding metadata form components

TylerZeroMaster committed:
Procedurally generate the metadata form.

Instead of hardcoding the jsx, use loops that dynamically create the form from a
list of inputs.

TylerZeroMaster committed:
Add validation, LO, APLO, and moduleId

sparksam commented on README.md:
```diff
@@ -38,15 +38,27 @@ will open a new VSCode window with the extension running. Clinking on a
+- [ ] Custom content fs implementation that can 'hide' private solutions/hints -
+      Maybe different names (not content id)
 
 ## Author
```
@TylerZeroMaster You can remove this section or add the other developer or something like @openstax/ce-be

TylerZeroMaster committed:
Add placeholder prop to InputSet and LO

TylerZeroMaster committed:
Collapse props in Bloom and Dok

TylerZeroMaster committed:
Add Nickname and Time fields

TylerZeroMaster committed:
Finish adding support for saving/loading metadata

Still some work to do with transforming and validating values.

The remaining work will need to wait until the metadata is well defined.

TylerZeroMaster committed:
Update onSaveError to match usage.

TylerZeroMaster committed:
Add validation for saving and loading metadata

And show specific error messages

TylerZeroMaster committed:
:bug: Send contentid to metadata form's save method

When a new h5p is created, the contentid is 'new'.

In this case, we want to use the contentid send back from the server.

TylerZeroMaster committed:
Add some component tests

Also, I started treating component props as if they were immutable.

TylerZeroMaster committed:
Add checkbox to toggle public/private solutions

TylerZeroMaster committed:
Add final conditional fields to md form

Renamed Book and Blooms for more consistent naming.

TylerZeroMaster committed:
Use sequential content ids

TylerZeroMaster committed:
Use accordion to show/hide form

TylerZeroMaster committed:
Make editor wait until the libraries are ready

TylerZeroMaster committed:
:art: Add list of authors to README

Also add spinning icon to editor loading status text.

TylerZeroMaster committed:
Changes required for gitpod

TylerZeroMaster committed:
Update package metadata

TylerZeroMaster committed:
Change module id to module path

TylerZeroMaster committed:
Do not mutate props in Book input

TylerZeroMaster committed:
Add some tests for metadata form

TylerZeroMaster committed:
Remove unused RestUtils

TylerZeroMaster committed:
Rename language server

TylerZeroMaster committed:
Change module path to module id

TylerZeroMaster committed:
Add new books

TylerZeroMaster committed:
Update package metadata for client and server

TylerZeroMaster committed:
Add additional tests for metadata form

Test encode/decode

TylerZeroMaster committed:
Add new metadata fields for nursing books

TylerZeroMaster committed:
Add tests for conditional nursing fields

TylerZeroMaster committed:
Add more tests for metadata form

Almost done: just need to test select box value change.

TylerZeroMaster committed:
Update AACN to match requirements

The requirements were updated

Also, add tests to match requirements

TylerZeroMaster committed:
Add element id field

TylerZeroMaster committed:
Change port to 27149

TylerZeroMaster committed:
:bug: Content id of 0 causes strange behavior

TylerZeroMaster committed:
Add pagination to content list component

TylerZeroMaster committed:
Add simple search by title to content list

TylerZeroMaster committed:
Add richer search syntax and proper id sorting

TylerZeroMaster committed:
Rename FileContentStoreage to FileContentStorage

Also fix a bug in createContentId

TylerZeroMaster committed:
Add tests for OSStorage

sparksam commented on client/src/app/components/ContentListComponent.tsx:
```diff
@@ -35,15 +40,54 @@ export default class ContentList extends React.Component<{
+              content.mainLibrary.toLocaleLowerCase().includes(s.slice(4));
+          default:
+            return (content: IContentListEntry) =>
+              content.title.toLocaleLowerCase().includes(s);
```
Got an error in my local when the folder title is a number. So had to add a cast like this. 
`String(content.title).toLocaleLowerCase().includes(s);`

TylerZeroMaster replied:
I think this may be a bug in the underlying h5p libraries. When you demonstrated the problem, we saw that the title was a number. In the underlying h5p library, `title` is [defined](https://github.com/Lumieducation/H5P-Nodejs-library/blob/10ff6134812f896b70b526685f52f42131d7fbbc/packages/h5p-server/src/types.ts#L118) as being a string. When the JSON is [loaded](https://github.com/Lumieducation/H5P-Nodejs-library/blob/10ff6134812f896b70b526685f52f42131d7fbbc/packages/h5p-server/src/implementation/fs/FileContentStorage.ts#L335), however, there is no type conversion or type checking.

I can override this in our file storage implementation and add type conversion and checking so that the types will match what is defined.

TylerZeroMaster replied:
In the process of fixing this, I might have found another bug in the h5p libraries. In `FileContentStorage.listContent`, each directory in the content directory is checked to ensure it has a `h5p.json` file in it. These checks are all done in parallel. On systems that have more strict limitations on the number of file descriptors that can be opened at once, this will cause an error when the number of directories checked exceeds the file descriptor limit.

Since we all use MacOS/Linux/Gitpod, this should not be a problem because these systems default to unlimited file descriptors. I think it would only be a problem on Windows. Consequently, it is probably not worth fixing now. I just wanted to add a note here incase it becomes a problem in the future so it might be easier to track down.

sparksam replied:
@TylerZeroMaster Please can you create a card that we can use to track this issue if it occurs in the future?

sparksam replied:
@TylerZeroMaster I am in for adding the type conversion and checks in our file storage implementation.

TylerZeroMaster committed:
:bug: coerce searchable h5p.json fields to string

TylerZeroMaster committed:
Add test for type conversion in getMetadata

TylerZeroMaster committed:
Add some tests for openstax h5p server

ckline-tryptic committed:
separate private answers

ckline-tryptic committed:
fix test for 'FileContentStorage'

TylerZeroMaster committed:
Add tests for utils.ts and fix some formatting.

TylerZeroMaster committed:
Do not delete extra information from metadata

TylerZeroMaster committed:
Add test case for saving metadata.

Make sure it updates the values as expected.

TylerZeroMaster committed:
Add error message for errors in server startup

TylerZeroMaster committed:
Improve test coverage reporting

- Coverage from client and server should be merged after tests complete
- Coverage should be reported in html format

TylerZeroMaster committed:
Update tests for FileContentStorage

TylerZeroMaster committed:
Test adding/subtracting inputs in InputSet

TylerZeroMaster committed:
Add test for debounce function

TylerZeroMaster committed:
Move encoders and decoders into one space

Also fix some typing

TylerZeroMaster committed:
Use nyc instead of istanbul for coverage reporting

istanbul is no longer maintained

TylerZeroMaster committed:
Add support for private H5P.Blanks answers

TylerZeroMaster committed:
Improve metadata form tests

TylerZeroMaster committed:
Add tests for private solutions

TylerZeroMaster committed:
Check the private solutions are saved as expected

TylerZeroMaster committed:
Change metadata keys to kabob casing

TylerZeroMaster committed:
:art: Lint

TylerZeroMaster committed:
Limit editor to supported libraries

TylerZeroMaster committed:
Exclude more from packaged extension

TylerZeroMaster committed:
More lint

TylerZeroMaster committed:
Package h5p libraries into extension.

Automatically downloads and gzips libraries after build

TODO:
- Get previous version of the libraries and include them in the final tar

TylerZeroMaster committed:
Compare library versions while updating

Instead of basing it off of the ctime of the tar file.

TylerZeroMaster committed:
Add support for custom archive file in download

TylerZeroMaster committed:
Add TrueFalse to list of supported libraries

TylerZeroMaster committed:
Move config into OpenStax directory...

since it contains details specific to our implementation.

TylerZeroMaster committed:
Make answer yankers/unyankers more extensible.

TylerZeroMaster committed:
Minor fixes to h5p library download script

- Throw err directly
- Be more specific about which libraries are being updated

TylerZeroMaster committed:
Store minimum private data for question sets

Question set merge deeply merges private data into questions.

TylerZeroMaster committed:
Make file content storage more ACID

TylerZeroMaster committed:
Remove orphaned content on error.

TylerZeroMaster committed:
Add tests for edge cases in file content storage

TylerZeroMaster committed:
Update README with info about included h5p libs

TylerZeroMaster committed:
Update packages

TylerZeroMaster committed:
Fix unhandled errors in base h5p server.

TylerZeroMaster committed:
:art: format README

TylerZeroMaster committed:
Test removing and combining private solutions

TylerZeroMaster committed:
Formatting mistakes

TylerZeroMaster committed:
Merge element id into module id & use kebab case

TylerZeroMaster committed:
Fix some formatting and typing issues.

TylerZeroMaster committed:
Store and input of book-specific values by book

Still WIP: may contain bugs :bug:

Still needs:
- validation of input values
- Storing individual values as individual values instead of arrays
- styling of book sub forms
- tests

TylerZeroMaster committed:
Update book input style to match mockup

TylerZeroMaster committed:
Move assertValue and assertType into utils

Add InputSetHandlerProps

TylerZeroMaster committed:
Add border to book fields

TylerZeroMaster committed:
Move element definitions out of render function

This also allows more information to be tied to element definitions.

TylerZeroMaster committed:
Save individual values for books individually

Rather than in arrays

TylerZeroMaster committed:
Fixup: use correct type

TylerZeroMaster committed:
Add validation for book values

Still need to check if at least one book is selected.

TylerZeroMaster committed:
Update validation for ap-lo

TylerZeroMaster committed:
Make bookInputHandlerProps pure

TylerZeroMaster committed:
Fixup: fix type of value used by decoders

TylerZeroMaster committed:
Update tests to support new book inputs

TylerZeroMaster committed:
Test encoding and decoding of book values

TylerZeroMaster committed:
Add tests for book input validation

TylerZeroMaster committed:
Add tests for adding/removing books

TylerZeroMaster committed:
Add tests for select boxes

...and fix formatting

TylerZeroMaster committed:
Fixup: fix typing

TylerZeroMaster committed:
Use full names for field titles

Except nclex (too long)

TylerZeroMaster committed:
Exclude empty values when encoding metadata

TylerZeroMaster committed:
Use grayscale banding for books

TylerZeroMaster committed:
Add tests for modifying book inputs and books

TylerZeroMaster committed:
Improve metadata form tests

- Make sure book values remain the same when adding/removing books
- Move some selectors to top-level constants
- Only do one thing per act()

TylerZeroMaster committed:
Add support for semantic overrides

Also add math tag to things that support math

TylerZeroMaster committed:
Add static router for h5p editor customizations

TylerZeroMaster committed:
Add patched ckeditor_wiris to build output

It is MIT licensed and the src is included

TylerZeroMaster committed:
Update plugin patch to always use modified indexOf

When working with multichoice questions, the MathType plugin starts
multiple times. We need it to use the patched String.indexOf each time.

TylerZeroMaster committed:
Add math tags in alterLibrarySemantics

Adding tags in the plugin activation script was not working.

TylerZeroMaster committed:
Enable math for MultiChoice

TylerZeroMaster committed:
Factory functions should not be properties.

This seemed like it could create confusion in the future if someone
expected these properties to return an existing router or create new.

TylerZeroMaster committed:
:bug: content id collisions

They were being sorted as strings rather than integers, ex: 1, 100, 2, etc.

In this case, the ids generated would have been 1, 2, 2, .. until 100
was reached.

TylerZeroMaster committed:
Override window.open for MathType

MathType uses this method for showing math source.
Instead of opening a window, show the content as html.

TylerZeroMaster committed:
Add support for rendering math in the h5p player.

TylerZeroMaster committed:
Add title bar with close button to html window

TylerZeroMaster committed:
Update dependencies

TylerZeroMaster committed:
Update tests to reflect changes to config class

TylerZeroMaster committed:
Add books.xml parsing to server

TylerZeroMaster committed:
Use values from books.xml to create config

Or do not: there are fallback values that are used if the books.xml does
not exist, but maybe this should be an error in the future.

TylerZeroMaster committed:
Test that alterLibrarySemantics does not mutate

It does (oops)

TylerZeroMaster committed:
Make alterLibrarySemantics pure

TylerZeroMaster committed:
Add visual indication of required fields

TylerZeroMaster committed:
Open first accordion item by default

TylerZeroMaster committed:
During tests, only reveal the form if it is hidden

TylerZeroMaster committed:
Do not allow duplicate titles

Add an error when an h5p with the supplied title already exists

Improve error handling so that the error message is shown to the user

Only show errors that extend CustomBaseError to the user

TylerZeroMaster committed:
Lint

TylerZeroMaster committed:
Title is only duplicated if the id does not match

If you are editing an h5p, then you should be allowed to save with the
same name as long as the id matches

TylerZeroMaster committed:
Be more specific in errors sent to the client

Forward all error messages to client for content endpoints

TylerZeroMaster committed:
Support additional semantic overrides

TylerZeroMaster committed:
Make solutions public by default and test accordion

Make sure the accordion opens and closes correctly

TylerZeroMaster committed:
Use Jest's expect patterns for testing errors

TylerZeroMaster committed:
Add more test cases for semantic overrides

TylerZeroMaster committed:
Make nickname optional and default to h5p title

TylerZeroMaster committed:
Add editor title button to open h5p editor

TylerZeroMaster committed:
Make selects clearable if their value is optional

TylerZeroMaster committed:
Use largest content id + 1 to generate content id

TylerZeroMaster committed:
Delete private solutions when deleting content

TylerZeroMaster committed:
Make metadata nickname equal title

TylerZeroMaster committed:
Rename ckeditor-plugins to editor-plugins

TylerZeroMaster committed:
:art: fix formatting

TylerZeroMaster committed:
Remove nickname field

Combined into h5p title

TylerZeroMaster committed:
Update tests for FileContentStorage

- No nickname
- Generated ids = largest id + 1
- Only directories that have h5p files count towards largest id

TylerZeroMaster committed:
:bug: Book fields should be cleared with book

Making the book field clearable allowed book to be empty string.
Prior, book could only be a non-empty string or undefined.
The behavior should be the same for empty string as undefined.

TylerZeroMaster committed:
Fix minor issues

- Sort contentIds as strings rather than numbers
- Ensure new content always exists first
- Move to page 1 when new content is created
- Add required arguments for config constructor in downloadLibraries

TylerZeroMaster committed:
Save osMeta in FileContentStore.addContent

TylerZeroMaster committed:
Update tests

New snapshot in FileContentStorage is because the value was not saved to
disk this change

Snapshot change in OpenstaxMetadataForm is from snapshotting encoded
values rather than OpenstaxMetadataForm.save.calls

TylerZeroMaster committed:
Use random id

TylerZeroMaster committed:
Force remove privatePath incase it does not exist

TylerZeroMaster committed:
Add support for arbitrary* HTML in content.json

* HTML tags still need to be added to an allow list otherwise the
client will strip them out

This allow list currently lives in
server/src/models/OpenStax/H5PEditor.ts

It is named `supportedTags` and is added to libraries that
should support HTML (defined in config.ts)

TylerZeroMaster committed:
Created an error if a supported library is missing

TylerZeroMaster committed:
Update h5p dependencies

TylerZeroMaster committed:
Update client tests

Needed to add jest-environment to account for some problems with jsdom

TylerZeroMaster committed:
Fix bug where nickname would be saved

TylerZeroMaster committed:
Update server tests

TylerZeroMaster committed:
Add some additional explanation

Also update formatting in a few areas

TylerZeroMaster committed:
Update h5p editor tests: test monkey patch

And filtering of supported libraries

TylerZeroMaster committed:
Remove unused methods of FileContentStorage

TylerZeroMaster committed:
Add delete confirmation

TylerZeroMaster committed:
Lint server

TylerZeroMaster committed:
Lint custom parts of client

TylerZeroMaster committed:
Lint client: formatting and unused imports only

TylerZeroMaster committed:
Slightly more involved linting fixes

TylerZeroMaster committed:
Most involved linting fixes with new config too

TylerZeroMaster committed:
Add tests to github actions

TylerZeroMaster committed:
Audit update packages

TylerZeroMaster committed:
No implicit override or implicit any in client

TylerZeroMaster committed:
Disable library download during CI testing

The download is very unreliable

TylerZeroMaster committed:
Move common utils into a common utils file

TylerZeroMaster committed:
Explain how to add CKEditor plugins

philschatz committed:
include re-copying the static files

TylerZeroMaster committed:
Install one h5p library at a time; wait on error

TylerZeroMaster committed:
Add plugins for HTML editing within the H5P editor

TylerZeroMaster committed:
Update README with new plugin inclusion process

TylerZeroMaster committed:
Improve build process

- Rename downloadLibraries to postbuild
- Move copy-extras into postbuild script
- Use npm-run-all instead of &&
- Update README
- Fix problem with gitpod.yml

TylerZeroMaster committed:
Fix ckeditor skin

The skin that was being used lacked certain icons.

Copying this newer version's css and icons fixes the issue.

ckline-tryptic committed:
add new books and sort them

ckline-tryptic committed:
Merge pull request #5 from openstax/openstax-update-new-books

add new books and sort them

ckline-tryptic committed:
Merge pull request #6 from openstax/openstax-update-new-books

Openstax update new books

ckline-tryptic committed:
fix book sorting

TylerZeroMaster committed:
Add image h5pimageupload plugin

TylerZeroMaster committed:
Fix memory leak in mergeH5PIntegration

Script and style arrays were being duplicated

TylerZeroMaster committed:
Improve FileContentStorage

Try to construct paths in as few places as possible

TylerZeroMaster committed:
Add support for images in content

Scan content properties for HTML image tags

Replace temporary paths with permanent paths

Name images based on filename data attribute

Add base url to player, editor, and ckeditor so images load

Look for files in both public and private directories

TylerZeroMaster committed:
Check for remaining temp paths before saving and...

improve a few things in ContentMutators

TylerZeroMaster committed:
Use new metadata schema

- Use more strongly typed metadata models
- Make conversion from form state to saved state less procedural
- Add CollaboratorSolutions
- Change module-id to context

TylerZeroMaster committed:
Add attachments to metadata

List all image paths relative to content.json

TylerZeroMaster committed:
Use media directory instead of images

Improve parsing of temporary file name

TylerZeroMaster committed:
Make optional fields optional and add placeholders

TylerZeroMaster committed:
Move randomId back into client

This was mistakenly moved: the way it is designed, it will only work on the client

Also limit errata id to 7 characters

TylerZeroMaster committed:
Update tests for FileContentStorage

TylerZeroMaster committed:
Fix bug: check for empty string when getting attributes

TylerZeroMaster committed:
Update tests for client

- Fix context pattern
- never return undefined from getIdx in bookInputSetHandlerProps

TylerZeroMaster committed:
Add additional tests for FileContentStorage

Check that public and private paths are searched

TylerZeroMaster committed:
Update README and client jest config

TylerZeroMaster committed:
Format README

TylerZeroMaster committed:
Move attachments when their visibility changes

TylerZeroMaster committed:
Include attachments from collaborator solutions

TylerZeroMaster committed:
Use labels instead of headers

TylerZeroMaster committed:
Indicate save errors more explicitly

Make the save button turn red and have a different icon

TylerZeroMaster committed:
Add error messages to value assertions in FileContentStorage

TylerZeroMaster committed:
Use proxy to override semantic entry properties

The overrides modify values in-place. They are sent a deep copy of
semantic entries.

This is a simpler and more flexible way to perform property overrides.

Limit the question types for question sets to MultiChoice, Blanks,
and TrueFalse

TylerZeroMaster committed:
Add missing html field name: 'tip'

It would be better to use semantics and check if the `widget` is html and `type` is text, but that is a little awkward.

TylerZeroMaster committed:
Set default dimensions for inserted images

TylerZeroMaster committed:
Only set image dimensions to default once

Was setting image dimensions to default when the width/height were manually changed

Caused by load event happening multiple times

TylerZeroMaster committed:
Parse html content as xhtml instead of xml

Fixes parse error caused by &nbsp; being inserted on empty lines

TylerZeroMaster committed:
Parse all content strings as HTML instead of a subset

Should prevent errors related to temporary images not being replaced

Example: 'ERROR: Found unexpected temp path in ...'

TylerZeroMaster committed:
Further simplify alterLibrarySemantics

Add unit tests for all alterations

TylerZeroMaster committed:
Add button to toggle preview for CollaboratorSolution

It does render mathml  when the preview is toggled

TylerZeroMaster committed:
Add tests for confirmation-dialog and collaborator-solutions

ckline-tryptic committed:
add attachments from metadata
eliminate duplicate attachments

ckline-tryptic committed:
delete unreferenced attachments

ckline-tryptic committed:
test removal of unreferenced images

ckline-tryptic committed:
update snapshot

TylerZeroMaster committed:
H5P php submodules (#8)

* Add h5p-php submodules

* Include H5P libraries directly in the extension

...instead of downloading them at runtime

* Remove unused downloadLibraries

* Add adjustable verbosity to extractArchive

...and make it easier to use.

* Use single quotes instead of ``

* Improve postbuild

- Use path.resolve
- Move library installation into a separate function

* Pin h5p submodules to 1.24.4

* Include patch for H5P editor to enable validation

* Update postbuild.ts

- Unwind paths
- Add pre-flight with helpful error messages
- Add integrity checks
- Further separate concerns
- Make everything more declarative

* Postbuild should not download h5p libraries during CI tests

* Define cross-referenced path names in config

- These names can change; you should only need to change them in one place
- Stop creating H5P directories in preflight (archive step catches missing paths)
- Update createH5PServer to use names defined in config

* Improve error handling in postbuild

Clarify some errors thrown by underlying libraries

* Exclude libraries from archive when CI_TEST is set

TylerZeroMaster committed:
Update documentation for building and running the extension (#9)

* Update documentation for building and running the extension

- Remove watch command from gitpod yml

* Update .gitpod.yml

Use package instead of build

Co-authored-by: Philip Schatz <253202+philschatz@users.noreply.github.com>

---------

Co-authored-by: Philip Schatz <253202+philschatz@users.noreply.github.com>

TylerZeroMaster committed:
2152 cat release (#10)

* Fix metadata

* Add scripts for building and publishing

* Remove unused ARCHIVE_FILE argument from postbuild

* npm audit fix

* Update README and add CLEAN_H5P environment variable

When this variable is set, the build script will not try to download the previous version of the extension.

* Fix formatting

TylerZeroMaster committed:
Remove attachments after all have been copied (#11)

* Remove attachments after all have been copied

Removing them as they were being copied was error prone

* Add test case

Fix some typos in comments

TylerZeroMaster committed:
Add misc documentation and set version to 1.0.0

TylerZeroMaster committed:
Add license for bootstrap skin

TylerZeroMaster committed:
Fix README formatting
